packages/core/src/tracing/instrumentation/messaging/rdkafka.js

```diff
@@ -0,0 +1,417 @@
+            }
+
+            // Only fall back to legacy binary trace correlation headers if no new header is present.
+            if (traceId == null && parentSpanId == null && level == null) {
+              // The newer string header format has not been found, fall back to legacy binary headers.
+              if (instanaHeadersAsObject[constants.kafkaTraceContextHeaderNameBinary]) {
+                const traceContextBuffer = instanaHeadersAsObject[constants.kafkaTraceContextHeaderNameBinary];
+                if (Buffer.isBuffer(traceContextBuffer) && traceContextBuffer.length === 24) {
+                  const traceContext = tracingUtil.readTraceContextFromBuffer(traceContextBuffer);
+                  traceId = traceContext.t;
+                  parentSpanId = traceContext.s;
+                }
+              }
+              level = readTraceLevelBinary(instanaHeadersAsObject);
+            }
+            removeInstanaHeadersFromMessage(messageData);
+          }
+
+          if (event === 'data' || event === 'error') {
+            cls.ns.runAndReturn(function () {
+              if (level !== '1') {
+                cls.setTracingLevel(level || '0');
+                return originalEmit.apply(ctx, originalArgs);
+              }
+              const span = cls.startSpan('kafka', constants.ENTRY, traceId, parentSpanId);
+              span.stack = tracingUtil.getStackTrace(instrumentedEmit, 1);
+              span.data.kafka = {
+                access: 'consume',
+                service: messageData.topic
+              };
+
+              if (event === 'error') {
+                delete messageData.headers;
+                span.ec = 1;
+                span.data.kafka.error = messageData.message;
+              }
+
+              setImmediate(() => {
+                span.d = Date.now() - span.ts;
+                span.transmit();
+              });
+
+              return originalEmit.apply(ctx, originalArgs);
+            });
+          } else {
+            return originalEmit.apply(ctx, originalArgs);
+          }
+        }
+      } else {
+        return originalEmit.apply(ctx, originalArgs);
+      }
+    });
+  } else {
+    return originalEmit.apply(ctx, originalArgs);
+  }
+}
+
+function readTraceLevelBinary(instanaHeadersAsObject) {
+  if (instanaHeadersAsObject[constants.kafkaTraceLevelHeaderNameBinary]) {
+    const traceLevelBuffer = instanaHeadersAsObject[constants.kafkaTraceLevelHeaderNameBinary];
+    if (Buffer.isBuffer(traceLevelBuffer) && traceLevelBuffer.length >= 1) {
+      return String(traceLevelBuffer.readInt8());
+    }
+  }
+  return '1';
+}
+
+function addTraceContextHeader(headers, span) {
+  if (!traceCorrelationEnabled) {
+    return headers;
+  }
+  switch (headerFormat) {
+    case 'binary':
+      return addTraceContextHeaderBinary(headers, span);
+    case 'string':
+      return addTraceContextHeaderString(headers, span);
+    case 'both':
+    // fall through (both is the default)
```

Comment:
Question: Do we actually want to use `both` as default even though we know the binary headers will not be handled correctly? WDYT about defaulting to `string` directly as long as the issue in node-rdkafka is not fixed? 

OTOH, if the node-rdkafka issue gets fixed, we get working trace correlation automatically, even if the consumer on the other end only understands the binary Instana headers. Maybe `both` is actually the correct choice. 🤔

Reply:
Trying to think from a customer's perspective, I think keeping "both" sounds correct.
As you said, if node-rdkafka is fixed, the default option of "both" becomes true again.
And we have the node-rdkafka current limitation already documented in the docs.

My 2 cents: keep it as is. 
If you folks decide to change it, we just need some sort of reminder to change it back when node-rdkafka is updated with the fix. And even then, there may be the need of a version validation to switch the default behavior.

Reply:
Yes, that makes sense. Thanks for pointing this out.

Reply:
Summary: 

- currently is "binary"
- we want to switch to "both" for rdkafka because customer does not need to set "headerFormat" explicitly
- "headerFormat" is not even documented yet
- using "both" does not hurt, because we just send two versions of the header format